storage/flux/table.gen.go

```diff
@@ -123,47 +127,128 @@ func newFloatWindowTable(
 			cur:   cur,
 		},
 		windowEvery: every,
+		createEmpty: createEmpty,
+	}
+	if t.createEmpty {
+		start := int64(bounds.Start)
+		t.nextTS = start + (every - start%every)
 	}
 	t.readTags(tags)
 	t.advance()
 
 	return t
 }
 
-func (t *floatWindowTable) advance() bool {
+func (t *floatWindowTable) Do(f func(flux.ColReader) error) error {
+	return t.do(f, t.advance)
+}
+
+// createNextWindow will read the timestamps from the array
+// cursor and construct the values for the next window.
+func (t *floatWindowTable) createNextWindow() (start, stop *array.Int64, ok bool) {
+	var stopT int64
+	if t.createEmpty {
+		stopT = t.nextTS
+		t.nextTS += t.windowEvery
+	} else {
+		if !t.nextBuffer() {
+			return nil, nil, false
+		}
+		stopT = t.arr.Timestamps[t.idxInArr]
+	}
+
+	// Regain the window start time from the window end time.
+	startT := stopT - t.windowEvery
+	if startT < int64(t.bounds.Start) {
+		startT = int64(t.bounds.Start)
+	}
+	if stopT > int64(t.bounds.Stop) {
+		stopT = int64(t.bounds.Stop)
+	}
+
+	// If the start time is after our stop boundary,
+	// we exit here when create empty is true.
+	if t.createEmpty && startT >= int64(t.bounds.Stop) {
+		return nil, nil, false
+	}
+	start = arrow.NewInt([]int64{startT}, t.alloc)
+	stop = arrow.NewInt([]int64{stopT}, t.alloc)
+	return start, stop, true
+}
+
+// nextAt will retrieve the next value that can be used with
+// the given stop timestamp. If no values can be used with the timestamp,
+// it will return the default value and false.
+func (t *floatWindowTable) nextAt(ts int64) (v float64, ok bool) {
+	if !t.nextBuffer() {
+		return
+	} else if !t.isInWindow(ts, t.arr.Timestamps[t.idxInArr]) {
+		return
+	}
+	v, ok = t.arr.Values[t.idxInArr], true
+	t.idxInArr++
+	return v, ok
+}
+
+// isInWindow will check if the given time at stop can be used within
+// the window stop time for ts. The ts may be a truncated stop time
+// because of a restricted boundary while stop will be the true
+// stop time returned by storage.
+func (t *floatWindowTable) isInWindow(ts int64, stop int64) bool {
+	start := stop - t.windowEvery
+	return start < ts && ts <= stop
```

Comment:
I think this should be `start <= ts && ts < stop`?
Given a timestamp `t` and window parameter `every`, the start of the window that `t` belongs to is:
```
start = t - t % every
```
(See: https://github.com/influxdata/flux/blob/master/execute/window.go#L110 and https://github.com/influxdata/influxdb/blob/master/storage/reads/array_cursor.gen.go#L254)

Then we can know:
```
start % every == 0, start - start % every == start
(start + every) - (start + every) % every == start + every
```

Therefore, the window range should include the `start` but exclude the `stop`.

Reply:
This was really confusing for me but I think the reason is because we are looking at the stop value. It might be worth a comment. When a window has a stop value of 30 seconds, that means it is covering the window of `[20, 30)` which doesn't actually include 30. So it appears like this should be `start <= ts && ts < stop`, but we're matching stop values for the interval. For the window `[20, 30)`, the stop value can be anywhere from `(20, 30]`. It's impossible for the stop value to be `20` exactly because then it would be the stop value for `[10, 20)`.